gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -0,0 +1,598 @@
+
+  private final String appMasterJvmArgs;
+
+  // Yarn application ID
+  private volatile Optional<ApplicationId> applicationId = Optional.absent();
+
+  // This flag tells if the Yarn application has already completed. This is used to
+  // tell if it is necessary to send a shutdown message to the ApplicationMaster.
+  private volatile boolean applicationCompleted = false;
+
+  private volatile boolean stopped = false;
+
+  public GobblinYarnAppLauncher(Config config) throws IOException {
+    this.config = config;
+
+    this.appName = config.getString(ConfigurationConstants.APPLICATION_NAME_KEY);
+    this.appQueueName = config.getString(ConfigurationConstants.APP_QUEUE_KEY);
+
+    String zkConnectionString = config.getString(ConfigurationConstants.ZK_CONNECTION_STRING_KEY);
+    LOGGER.info("Using ZooKeeper connection string: " + zkConnectionString);
+
+    this.helixManager = HelixManagerFactory.getZKHelixManager(
+        config.getString(ConfigurationConstants.HELIX_CLUSTER_NAME_KEY), YarnHelixUtils.getHostname(),
+        InstanceType.SPECTATOR, zkConnectionString);
+
+    this.yarnConfiguration = new YarnConfiguration();
+    this.yarnConfiguration.set("fs.automatic.close", "false");
+    this.yarnClient = YarnClient.createYarnClient();
+    this.yarnClient.init(this.yarnConfiguration);
+
+    this.fs = config.hasPath(ConfigurationKeys.FS_URI_KEY) ?
+        FileSystem.get(URI.create(config.getString(ConfigurationKeys.FS_URI_KEY)), this.yarnConfiguration) :
+        FileSystem.get(this.yarnConfiguration);
+
+    List<Service> services = Lists.newArrayList();
+    if (UserGroupInformation.isSecurityEnabled()) {
+      LOGGER.info("Adding YarnAppSecurityManager since security is enabled");
+      services.add(new YarnAppSecurityManager(config, this.helixManager, this.fs));
+    }
+    this.serviceManager = new ServiceManager(services);
+
+    this.applicationStatusMonitor = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor(
+        ExecutorsUtils.newThreadFactory(Optional.of(LOGGER), Optional.of("GobblinYarnAppStatusMonitor"))));
+
+    this.appMasterJvmArgs = Strings.nullToEmpty(config.getString(ConfigurationConstants.APP_MASTER_JVM_ARGS_KEY));
+  }
+
+  /**
+   * Launch a new Gobblin instance on Yarn.
+   *
+   * @throws IOException if there's something wrong launching the application
+   * @throws YarnException if there's something wrong launching the application
+   */
+  public void launch() throws IOException, YarnException {
+    createGobblinYarnHelixCluster();
+
+    try {
+      this.helixManager.connect();
+    } catch (Exception e) {
+      LOGGER.error("HelixManager failed to connect", e);
+      throw Throwables.propagate(e);
+    }
+
+    // Start all the services running in the ApplicationMaster
+    this.serviceManager.startAsync();
+
+    this.yarnClient.start();
+    this.applicationId = Optional.of(setupAndSubmitApplication());
+
+    ListenableFuture<ApplicationReport> appReportFuture =
```

Comment:
Would it be simpler to just have a regular `Executors.newSingleThreadScheduledExecutor` that is scheduled the below `Callable` to run every 60 seconds. The `call()` method can check if the `appState` is `FINISHED`, `FAILED`, or `KILLED`; if it is then it executes `GobblinYarnAppLauncher.this.stop()` and sets `applicationCompleted = true`. Otherwise it does nothing.

If an exception is ever thrown, it should be accessible via a regular `Future` from the thrown `ExecutionException`.